### PR TITLE
[zoneminder] Replacement for ZoneMinder binding

### PR DESCRIPTION
**Pull request by mhilbush**
This PR is a replacement for PR #7412, which was submitted against the 2.5.x branch. The 2.5.x version of the binding has been reviewed by @cpmeister and @fwolter. There are still a couple recent changes that are awaiting review by @fwolter .

**Background**
The current OH2 ZoneMinder binding (org.openhab.binding.zoneminder) is unmaintained.
It also has a number of shortcomings:

- There are 6266 SLOC (as counted by SLOCCount). This count includes the zoneminder binding itself, as well as the zoneminder4j library that the binding uses.
- The binding was written for older versions of Zoneminder. Since then, the Zoneminder API has evolved considerably, such that much of this binding's functionality can now be accomplished much more easily with version 2.0 of the Zoneminder API.
- The current binding doesn't support Zoneminder's latest JWT-based authentication scheme, which improves security over the previous authentication methods. Anyone who has enabled authentication in recent versions of Zoneminder is unable to use the old binding.
- The current binding (and the library it uses) do not use null annotations. Adding null annotations is a significant amount of work.
- There is an abandoned 2-year old PR #3583 that added 4429 LOC and deleted 1649 LOC. That PR was closed a year ago. I took a look at the PR's review comments, as well as the code. Resurrecting it would be a monumental amount of work.

**Implications**
I wrote a new version of the binding using ZoneMinder version 1.34.0 and Zoneminder API version 2.0 as a base. My binding is 1698 SLOC (as counted by SLOCCount) versus the current binding's 6266 SLOC. It replaces the old binding in its entirety. It also is not backwards compatible with ZoneMinder versions prior to 1.34.0. Users of the old binding who upgrade to OH3 will need to upgrade their ZoneMinder to use this new binding.

The new binding has been in use for quite a while in 2.5.x. See community post here (https://community.openhab.org/t/new-zoneminder-binding-for-zoneminder-versions-1-34-0/95112).

Signed-off-by: Mark Hilbush <mark@hilbush.com>

### CODE REVIEW

**Comment by jeremystucki** [comment]
Is there a way to keep the commit history? (it says that the main branch is unrelated to the 2.5.x branch)

**Comment by Hilbrand** [comment]
You'll have a conflict on the `.classpath` because that file has been removed from GitHub.

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java:236]: `refreshImage()` is already called asynchronously, why does `getImage()` need to be asynchronous then?

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java:73]:
```suggestion
    private final Map<String, State> monitorStatusCache = new ConcurrentHashMap<>();
```

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/config/config.xml
Comment (cpmeister) [inline bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/config/config.xml:84]: If it isn't required what is the default value?
Also shouldn't this have a minimum value?

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml
Comment (cpmeister) [inline bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:202]: I'm not sure that you can have fractional frames.
```suggestion
		<state readOnly="true" pattern="%d"></state>
```

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml
Comment (cpmeister) [inline bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:208]: same here
```suggestion
		<state readOnly="true" pattern="%d"></state>
```

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml
Comment (cpmeister) [inline bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:84]: Should this be a thing property instead?

**Review by cpmeister — CHANGES_REQUESTED** [review]
Looks pretty good, just have a few questions.

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmAuth.java
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmAuth.java:134]: No, no. I need to really make sure it's null. lol

Fixed.

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmAuth.java
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmAuth.java:40]: Done

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java:73]: Done

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/config/config.xml
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/config/config.xml:84]: No, the default is to not refresh the image, in which case you leave the field blank.

Yes, I'll add a minimum.

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:202]: Correct. Fixed.

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:208]: Fixed

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:84]: If you change the monitor id in Zoneminder, you should be able to change it here as well. Otherwise, you would need to delete and readd the thing.

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/discovery/MonitorDiscoveryService.java
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/discovery/MonitorDiscoveryService.java:56]: I don't know. It's done this way in a number of other bindings. And, I don't think we know the bridge handler in activate.

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java:236]: My original thinking was that I didn't want `refreshImage` to block waiting for the image. But, I can't think of a reason why that would be necessary. I've removed it.

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmBridgeHandler.java
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmBridgeHandler.java:445]: Agreed. Done.

File: bundles/org.openhab.binding.zoneminder/README.md
Comment (cpmeister) [inline bundles/org.openhab.binding.zoneminder/README.md:56]: The config xml specifies that these are not required.

File: bundles/org.openhab.binding.zoneminder/README.md
Comment (cpmeister) [inline bundles/org.openhab.binding.zoneminder/README.md:52]: This one too

**Review by cpmeister — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.zoneminder/README.md
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/README.md:56]: Fixed

File: bundles/org.openhab.binding.zoneminder/README.md
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/README.md:52]: Also fixed.

**Review by cpmeister — APPROVED** [review]
LGTM

**Comment by cpmeister** [comment]
Looks like you need to run spotless on the config.xml

**Comment by mhilbush** [comment]
Yeah, I was just looking at that. I didn't think I made a change that would've tweaked the formatting. I'll fix now.

**Comment by mhilbush** [comment]
Ok, the build issue should be fixed.

But, I need to look at this.
```
[WARNING] Code Analysis Tool has found:
 0 error(s)!
 2 warning(s)
 0 info(s)
[WARNING] OH-INF.config.config.xml:[0]
Unused configuration reference with uri - thing-type:zoneminder:monitor
[WARNING] OH-INF.config.config.xml:[0]
Unused configuration reference with uri - thing-type:zoneminder:server
```

**Comment by mhilbush** [comment]
I don't understand what's causing the above warnings. I've never seen these before. In testing the binding, it's handling the config properly.

**Comment by mhilbush** [comment]
And now I need to fix the merge conflict with CODEOWNERS...

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java:112]: This could be debug or warn.

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/config/config.xml
Comment (fwolter) [inline bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/config/config.xml:37]:
```suggestion
		<parameter name="defaultImageRefreshInterval" type="integer" unit="s" required="false" groupName="config-info">
```

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [reply bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:84]: How would a user do this, as it is read only? Isn't this a configuration parameter, either?

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.zoneminder/README.md
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/README.md:50]: Sorry @fwolter. I had responded to your question in the old PR, then closed the PR before you had a chance to see my response.

Here's what I said in the original PR. Perhaps you have a suggestion for how to do it differently.

> This is how I wanted to do it originally. However, if you define the parameter to have a default, I don't believe it's then possible to leave the parameter empty. If you leave it empty, OH will just insert the default. As there are some Zoneminder installations that need the path to be empty, this was the only way I could think to do it.

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/action/ZmActions.java
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/action/ZmActions.java:75]: Since DSL rules will still be supported in OH3, I don't recall seeing anything that fixes the root cause of the issue. And the [issue in core](https://github.com/openhab/openhab-core/issues/1265) is still open. But @cweitkamp would know for sure. Let wait for his reply.

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:84]: Thank you for making me rethink this. I'll try to clarify, then we can figure out what to do.

- The above channel contains the monitor id returned by the Zoneminder API. I made it a channel because the user likely will want to include the monitor id in their sitemap or HABpanel UI. As the monitor id is returned by the Zoneminder API, and cannot be changed by the user, I believe it should be readonly.

- As for the config parameter, when creating a monitor thing manually, or using a `.things` file, the user needs to specify the monitor id as part of defining the thing. When the monitor thing is created through discovery, it is set by the binding. Also, the monitor id the representation property.

So, what to do? I'm not sure how to resolve this without impacting one of the two points above. WDYT?

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/action/ZmActions.java
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/action/ZmActions.java:88]: Some bindings use info, but most use warn. I'll make it warn.

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/handler/ZmMonitorHandler.java:112]: Probably should be warn.

File: bundles/org.openhab.binding.zoneminder/README.md
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/README.md:156]: Agreed. Merged.

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/config/config.xml
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/config/config.xml:37]: Fixed

**Comment by mhilbush** [comment]
@fwolter I addressed most of your feedback in the latest commit. For the remaining ones, I've added some comments and questions. Thanks for your help!

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [reply bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:84]: Your arguments are valid. If I see correctly, the State of the Channel `id` is updated if the user changes the config parameter `monitorId`. I think it's good to go.

File: bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/action/ZmActions.java
Comment (cweitkamp) [reply bundles/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/action/ZmActions.java:75]: Please correct me if I am wrong, I am not really up-to-date right now: IIRC there was no real confirmation for `ThingActions` are working correctly in OH3 NGRE and we wanted to keep those workarounds and bulk remove them later.

File: bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/src/main/resources/OH-INF/thing/thing-types.xml:84]: Perfect. Thanks!

File: bundles/org.openhab.binding.zoneminder/README.md
Comment (mhilbush) [reply bundles/org.openhab.binding.zoneminder/README.md:51]: So, yes, the default Zoneminder installation would be:
```
http://localhost/zm/path/to/api/endpoint
```

And an alternate would be:
```
http://localhost/path/to/api/endpoint
```

How obvious!!! Sometimes you can't see the forest for the trees. lol

I'll make the change.

**Comment by mhilbush** [comment]
I've reworked the config parameters and the documentation.

**Review by fwolter — APPROVED** [review]
LGTM

**Comment by mhilbush** [comment]
Thanks for your help on this @fwolter and @cpmeister!!

**Pull request merged**
